### PR TITLE
Added `data_format` to flatten layer.

### PR DESCRIPTION
**Pull request by joeyearsley**
If you create an inference graph in `NHWC` but have trained in `NCHW` you will realise that your inference predictions are different from train time.

This is due to the fact that the flatten layer does not respect channel orderings, therefore I have added a `data_format` keyword arg such that it is always reshaped to `NHWC` - as for dense layers, it does not matter.

### CODE REVIEW

**Comment by joeyearsley** [comment]
You are correct with the K.ndim, unsure if that will resolve the dynamic axis issue with CNTK.
If it doesn't, I will setup a CNTK docker image to debug that issue in.

**Comment by fchollet** [comment]
The issue persists with CNTK. @souptc could you please advise on how to fix it?

**Comment by souptc** [comment]
my bad, it is a bug in cntk_backend. In cntk_backend.py method "def permute_dimensions", line 1115:
`if num_dynamic_axis > 0 and pattern[:num_dynamic_axis] != current_layout[:num_dynamic_axis]:`

here current_layout is a tuple but pattern could be a list, so compare may failed un-expected...

Do you mind to fix this in this pr?

**Comment by joeyearsley** [comment]
Thanks @souptc! - that has fixed it.

File: keras/utils/test_utils.py [hunk 1]
```diff
@@ -116,6 +116,9 @@ def layer_test(layer_cls, kwargs={}, input_shape=None, input_dtype=None,
 
     # test as first layer in Sequential API
     layer_config = layer.get_config()
+    # deals with data_format in flatten
+    if 'data_format' in kwargs:
+        layer_config['data_format'] = kwargs['data_format']
```
Comment (fchollet) [inline keras/utils/test_utils.py]: Revert this change and add support for `data_format` in `get_config` for this layer

Comment (joeyearsley) [reply keras/utils/test_utils.py]: Forgot about that func, I've updated to reflect your comments.

**Comment by joeyearsley** [comment]
@fchollet Any more to do?

File: keras/layers/core.py [hunk 2]
```diff
@@ -465,6 +466,13 @@ def get_config(self):
 class Flatten(Layer):
     """Flattens the input. Does not affect the batch size.
 
+    Arguments:
```
[the remark this answers is not in the document]
Comment (joeyearsley) [reply keras/layers/core.py]: Done

**Comment by TimZaman** [comment]
I don't understand why we would always convert `channels_first` to `channels_last`? That seems suboptimal for `channels_last` models, for both training as well as inference.

**Comment by joeyearsley** [comment]
In what way? 

This is to make it easier to productise going from NCHW to NHWC. 

The dimension ordering of your 4d matrix matters before your dense layers, if you don’t account for these 2 formats then you get lots of regression tests breaking when you do the conversion for inference time.

Why is it suboptimal if it’s a flatten layer? - ignoring the one extra transpose added when using NCHW.

I don’t fully understand your final point.

**Comment by TimZaman** [comment]
It's suboptimal as it creates ops in the graph when training `channels_first` (more optimal for GPUs) that effectively don't do anything. Although I think we agree on that, your original intent here was good, but another change by Francois changed that.

I think this happened: with good intent, you created this initializer:
```patch
+   def __init__(self, data_format='channels_last', **kwargs):
```
the `data_format` is odd, as Keras usually sets this to `None`. The behavior with the `channels_last` here is good, because it's backwards compatible, and won't add any extraneous ops.

However, then this change happened https://github.com/keras-team/keras/commit/243338dc36c554676f1e9a5f733b445341d9f4f5 :
```patch
-    def __init__(self, data_format='channels_last', **kwargs):
+    def __init__(self, data_format=None, **kwargs):
```

Which suddenly is not backwards compatible anymore, and changes the behaviour of both training and doing inference on `channels_first` models!

**Comment by joeyearsley** [comment]
Ah, now I see the confusion! Yes I agree
